src/query/api/v1/handler/prometheus/tags/common.go

```diff
@@ -0,0 +1,154 @@
+// THE SOFTWARE.
+
+package tags
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/m3db/m3/src/query/errors"
+	"github.com/m3db/m3/src/query/models"
+	"github.com/m3db/m3/src/query/storage"
+	"github.com/m3db/m3/src/query/util/json"
+	"github.com/m3db/m3/src/x/net/http"
+)
+
+const (
+	queryParam          = "query"
+	filterNameTagsParam = "tag"
+	formatErrStr        = "error parsing param: %s, error: %v"
+)
+
+// parseParamsToQuery parses all params from the GET request
+func parseParamsToQuery(r *http.Request) (storage.CompleteTagsQuery, *xhttp.ParseError) {
+	tagQuery := storage.CompleteTagsQuery{}
+
+	query, err := parseQuery(r)
+	if err != nil {
+		return tagQuery, xhttp.NewParseError(fmt.Errorf(formatErrStr, queryParam, err), http.StatusBadRequest)
+	}
+
+	matchers, err := models.MatchersFromString(query)
+	if err != nil {
+		return tagQuery, xhttp.NewParseError(err, http.StatusBadRequest)
+	}
+
+	tagQuery.TagMatchers = matchers
+	// If there is a result type field present, parse it and set
+	// complete name only parameter appropriately. Otherwise, default
+	// to returning both completed tag names and values
+	if result := r.FormValue("result"); result != "" {
+		switch result {
+		case "default":
+			tagQuery.CompleteNameOnly = false
+		case "tagNamesOnly":
+			tagQuery.CompleteNameOnly = true
+		default:
+			return tagQuery, xhttp.NewParseError(errors.ErrInvalidResultParamError, http.StatusBadRequest)
+		}
+	}
+
+	filterNameTags := r.Form[filterNameTagsParam]
+	tagQuery.FilterNameTags = make([][]byte, len(filterNameTags))
+	for i, f := range filterNameTags {
+		tagQuery.FilterNameTags[i] = []byte(f)
+	}
+
+	return tagQuery, nil
+}
+
+func parseQuery(r *http.Request) (string, error) {
+	queries, ok := r.URL.Query()[queryParam]
+	if !ok || len(queries) == 0 || queries[0] == "" {
+		return "", errors.ErrNoQueryFound
+	}
+
+	// TODO: currently, we only support one target at a time
+	if len(queries) > 1 {
+		return "", errors.ErrBatchQuery
+	}
+
+	return queries[0], nil
+}
+
+func renderNameOnlyResultsJSON(
+	w io.Writer,
+	results []storage.CompletedTag,
+) error {
+	jw := json.NewWriter(w)
+	jw.BeginArray()
+
+	for _, tag := range results {
+		jw.WriteString(string(tag.Name))
+	}
+
+	jw.EndArray()
+
+	return jw.Close()
+}
+
+func renderDefaultResultsJSON(
+	w io.Writer,
+	results []storage.CompletedTag,
+) error {
+	jw := json.NewWriter(w)
+	jw.BeginObject()
+
+	jw.BeginObjectField("hits")
+	jw.WriteInt(len(results))
+
+	jw.BeginObjectField("tags")
+	jw.BeginArray()
+
+	for _, tag := range results {
+		jw.BeginObject()
+
+		jw.BeginObjectField("key")
+		jw.WriteString(string(tag.Name))
+
+		jw.BeginObjectField("values")
+		jw.BeginArray()
+		for _, value := range tag.Values {
+			jw.WriteString(string(value))
+		}
+		jw.EndArray()
+
+		jw.EndObject()
+	}
+	jw.EndArray()
+
+	jw.EndObject()
+
+	return jw.Close()
+}
+
+func renderResultsJSON(
+	w io.Writer,
+	result storage.CompleteTagsResult,
```

Comment:
Nit: looks like the result type is passed around as a pointer everywhere else; any reason to not follow the same convention here?